firedrake/dmhooks.py

```diff
@@ -64,11 +69,21 @@ def set_function_space(dm, V):
 
     .. note::
 
-       This stores a weakref to the function space in the DM, so you
-       should hold a strong reference somewhere else.
+       This stores the information necessary to make a function space given a DM.
 
     """
-    dm.setAttr("__fs_info__", weakref.ref(V))
+    mesh = V.mesh()
+
+    indices = []
+
+    Vtmp = V
+    while Vtmp.parent is not None:
+        indices.append(Vtmp.index)
+        Vtmp = Vtmp.parent
+    element = V.ufl_element()
+
+    info = (weakref.ref(mesh), element, tuple(reversed(indices)))
+    dm.setAttr("__fs_info__", info)
```

Comment:
FIxing, hopefully, problems to do with not finding functionspaces on DMs. We must have a mesh lying around, so store that and the necessary information to make the FS from it (instead of a weakref to the FS).